bridge/src/send_tx_with_receipt.rs

```diff
@@ -0,0 +1,368 @@
+                                    Some(receipt_block_number) => {
+                                        if last_block < receipt_block_number.as_u64() {
+                                            // transaction does not have enough confirmations
+                                            State::AwaitBlockNumber(transaction_hash)
+                                        } else {
+                                            return Ok(Async::Ready(receipt))
+                                        }
+                                    }
+                                }
+                            },
+                        }
+                    }
+                };
+
+                self.state = next_state;
+            }
+        }
+    }
+}
+
+enum State<T: Transport> {
+    AwaitBlockNumber {
+        future: Timeout<FromErr<CallFuture<U256, T::Out>, error::Error>>,
+        transaction: Option<TransactionRequest>,
+    },
+    AwaitReceipt(inner::SendTransactionWithReceipt<T>)
+}
+
+
+pub struct SendTransactionWithReceiptOptions<T> {
+    pub transport: T,
+    pub request_timeout: Duration,
+    pub poll_interval: Duration,
+    pub confirmations: u32,
+    pub transaction: TransactionRequest,
+}
+
+pub struct SendTransactionWithReceipt<T: Transport> {
+    request_timeout: Duration,
+    poll_interval: Duration,
+    transport: T,
+    state: State<T>,
+    confirmations: u32,
+}
+
+impl<T: Transport> SendTransactionWithReceipt<T> {
+    pub fn new(options: SendTransactionWithReceiptOptions<T>) -> Self {
+        let timer = Timer::default();
+
+        let future = web3::api::Eth::new(&options.transport).block_number();
+
+        let state = State::AwaitBlockNumber {
+            future: timer.timeout(future.from_err(), options.request_timeout),
+            transaction: Some(options.transaction),
+        };
+
+        SendTransactionWithReceipt {
+            request_timeout: options.request_timeout,
+            poll_interval: options.poll_interval,
+            transport: options.transport,
+            state,
+            confirmations: options.confirmations,
+        }
+    }
+}
+
+impl<T: Transport> Future for SendTransactionWithReceipt<T> {
+    type Item = TransactionReceipt;
+    type Error = error::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        loop {
+            let next_state = match self.state {
+                State::AwaitBlockNumber { ref mut future, ref mut transaction } => {
+                    let block_number = try_ready!(
+                        future
+                            .poll()
+                            .chain_err(|| "SendTransactionWithReceipt: fetching last block number failed")
+                    );
+                    info!("SendTransactionWithReceipt: got last block number {}", block_number);
+                    let transaction = transaction.take().expect("transaction should always be set");
```

Comment:
Should be more proofy - just mention that we are in `AwaitBlockNumber` state only initially and transaction is always `Some` initially.